src/Razor/benchmarks/Microsoft.AspNetCore.Razor.Microbenchmarks/LanguageServer/RazorSemanticTokensRangeEndpointBenchmark.cs

```diff
@@ -0,0 +1,179 @@
+
+    private DocumentVersionCache VersionCache { get; set; }
+
+    private Uri DocumentUri => DocumentContext.Uri;
+
+    private IDocumentSnapshot DocumentSnapshot => DocumentContext.Snapshot;
+
+    private VersionedDocumentContext DocumentContext { get; set; }
+
+    private Range Range { get; set; }
+
+    private new IRazorLogger Logger { get; set; }
+
+    private ProjectSnapshotManagerDispatcher ProjectSnapshotManagerDispatcher { get; set; }
+
+    private string PagesDirectory { get; set; }
+
+    private string ProjectFilePath { get; set; }
+
+    private string TargetPath { get; set; }
+
+    private CancellationToken CancellationToken { get; set; }
+
+    private RazorRequestContext RequestContext { get; set; }
+
+    private const int MaximumNumberOfCsSemanticRangesToReturn = 1000;
+
+    [Params(0, 100, MaximumNumberOfCsSemanticRangesToReturn)]
+    public int NumberOfCsSemanticRangesToReturn { get; set; }
+
+    private static List<SemanticRange> PregeneratedRandomSemanticRanges { get; set; }
+
+
+    [GlobalSetup]
+    public async Task InitializeRazorSemanticAsync()
+    {
+        EnsureServicesInitialized();
+
+        var projectRoot = Path.Combine(RepoRoot, "src", "Razor", "test", "testapps", "ComponentApp");
+        ProjectFilePath = Path.Combine(projectRoot, "ComponentApp.csproj");
+        PagesDirectory = Path.Combine(projectRoot, "Components", "Pages");
+        var filePath = Path.Combine(PagesDirectory, $"SemanticTokens.razor");
+        TargetPath = "/Components/Pages/SemanticTokens.razor";
+
+        var documentUri = new Uri(filePath);
+        var documentSnapshot = GetDocumentSnapshot(ProjectFilePath, filePath, TargetPath);
+        var version = 1;
+        DocumentContext = new VersionedDocumentContext(documentUri, documentSnapshot, version);
+        Logger = new NoopLogger();
+        SemanticTokensRangeEndpoint = new SemanticTokensRangeEndpoint();
+
+        var text = await DocumentContext.GetSourceTextAsync(CancellationToken.None).ConfigureAwait(false);
+        Range = new Range
+        {
+            Start = new Position { Line = 0, Character = 0 },
+            End = new Position { Line = text.Lines.Count - 1, Character = text.Lines.Last().Span.Length - 1 }
+        };
+
+        var documentVersion = 1;
+        CancellationToken = CancellationToken.None;
+        await UpdateDocumentAsync(documentVersion, DocumentSnapshot, CancellationToken).ConfigureAwait(false);
+
+        var languageServer = RazorLanguageServer.GetInnerLanguageServerForTesting();
+        RequestContext = new RazorRequestContext(DocumentContext, Logger, languageServer.GetLspServices());
+
+        var random = new Random();
+        var codeDocument = await DocumentContext.GetCodeDocumentAsync(CancellationToken);
+        PregeneratedRandomSemanticRanges = new List<SemanticRange>();
+        for (var i = 0; i < MaximumNumberOfCsSemanticRangesToReturn; i++)
```

Comment:
Does it not work to just use `NumberOfCsSemanticRangesToReturn` here?

Reply:
We can pre-gen all ranges and just take a subset for each test, instead of doing it each test

Reply:
Perhaps it depends what the benchmark results look like, but as a general rule I think its better to get as much setup work out of the benchmark itself, just to prevent it introducing noise to the results.